### PR TITLE
feat(server): adding support for EXAT PXAT option at set command

### PR DESCRIPTION
**Pull request by boazsade**
Signed-off-by: Boaz Sade <boaz@dragonflydb.io>

feat(server): support for the options EXAT and PXAT for the set command.
* For values less than 0 in the command returning "invalid expire time"
* For values that expired before the current time, the function return OK but will not set the value (per Redis today).
* For values in the future it would translate to milliseconds and do offset from current time (so we are keeping the offset).

### CODE REVIEW

File: src/server/string_family.cc [hunk 1]
```diff
@@ -464,7 +481,21 @@ void StringFamily::Set(CmdArgList args, ConnectionContext* cntx) {
...
+      if (int_arg <= 0) {
+        return builder->SendError(InvalidExpireTime("set"));
+      }
+      // for []AT we need to take expiration time as absolute from the value given
+      // check here and if the time is in the pass, return OK but don't set it
+      // Note that the time pass here for PXAT is in milliseconds, we must not change it!
+      if (cur_arg == "EXAT" || cur_arg == "PXAT") {
+        int_arg = CalculateAbsTime(int_arg, is_ms);
+        if (int_arg < 0) {
+          // this happened in the pass, just return, for some reason Redis reports OK in this case
```
Comment (romange) [inline src/server/string_family.cc:494]: pass -> past

File: src/server/string_family.cc [hunk 2]
```diff
@@ -464,7 +481,21 @@ void StringFamily::Set(CmdArgList args, ConnectionContext* cntx) {
         return builder->SendError(kInvalidIntErr);
       }
 
-      if (int_arg <= 0 || (!is_ms && int_arg >= kMaxExpireDeadlineSec)) {
+      // Since PXAT/EXAT can change this, we need to check this ahead
+      if (int_arg <= 0) {
+        return builder->SendError(InvalidExpireTime("set"));
+      }
+      // for []AT we need to take expiration time as absolute from the value given
+      // check here and if the time is in the pass, return OK but don't set it
```
Comment (romange) [inline src/server/string_family.cc:489]: pass -> past

File: src/server/string_family_test.cc [hunk 3]
```diff
@@ -355,4 +355,35 @@ TEST_F(StringFamilyTest, SetNx) {
...
+  resp = Run({"set", "foo", "bar", "PXAT", std::to_string(now - 23)});
+  ASSERT_THAT(resp, "OK");  // it would return OK but will not set the value (checked with Redis)
+  EXPECT_EQ(Run({"get", "foo"}).type, facade::RespExpr::NIL);
+
+  resp = Run({"set", "foo", "bar", "EXAT", std::to_string(time(nullptr) + 1)});
+  ASSERT_THAT(resp, "OK");  // valid expiration time
+  EXPECT_EQ(Run({"get", "foo"}), "bar");
+
+  resp = Run({"set", "foo2", "abc", "PXAT", std::to_string(now + 300)});
+  ASSERT_THAT(resp, "OK");  // it would return OK but will not set the value (checked with Redis)
```
Comment (romange) [inline src/server/string_family_test.cc:385]: you copied the wrong comment here

File: src/server/string_family.cc (hunk 1 above)
Comment (boazsade) [reply src/server/string_family.cc:494]: ok

File: src/server/string_family.cc (hunk 2 above)
Comment (boazsade) [reply src/server/string_family.cc:489]: ok

File: src/server/string_family_test.cc (hunk 3 above)
Comment (boazsade) [reply src/server/string_family_test.cc:385]: right - sorry

**Review by romange — APPROVED** [review]

**Pull request merged**